File: HSTracker/Logging/Handlers/AssetHandler.m
/*
 * This file is part of the HSTracker package.
 * (c) Benjamin Michotte <bmichotte@gmail.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 *
 * Created on 13/02/16.
 */
#import "AssetHandler.h"
#import "Game.h"

static NSString *const MedalRank = @"Medal_Ranked_(\\d+)";
static NSString *const UnloadingCard = @"unloading name=(\\w+_\\w+) family=CardPrefab persistent=False";

@implementation AssetHandler

+ (void)handle:(NSString *)line
{
  Game *game = [Game instance];
  if (game.awaitingRankedDetection) {
    game.lastAssetUnload = [NSDate date].timeIntervalSince1970;
    game.awaitingRankedDetection = NO;
  }

  if ([line isMatch:RX(MedalRank)]) {
    NSNumber *rank = @([[line firstMatch:RX(MedalRank)] integerValue]);
    [game setPlayerRank:rank];
  }
  else if ([line isMatch:RX(@"rank_window")]) {
    game.rankFound = YES;
    game.gameMode = EGameMode_Ranked;
  }
  else if ([line isMatch:RX(UnloadingCard)]) {
    RxMatch *match = [line firstMatchWithDetails:RX(UnloadingCard)];
    NSString *cardId = ((RxMatchGroup *) match.groups[1]).value;
    if (game.gameMode == EGameMode_Arena) {
      DDLogInfo(@"Possible arena card draft : %@ ?", cardId);
    }
    else {
      DDLogInfo(@"Possible constructed card draft : %@ ?", cardId);
    }
  }
  else if ([line isMatch:RX(@"unloading name=Tavern_Brawl")]) {
    game.gameMode = EGameMode_Brawl;
  }
}

@end

File: HSTracker/Logging/Handlers/AssetHandler.swift
/*
 * This file is part of the HSTracker package.
 * (c) Benjamin Michotte <bmichotte@gmail.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 *
 * Created on 13/02/16.
 */

import Foundation
import CleanroomLogger

struct AssetHandler {

    let UnloadingCard = "unloading name=(\\w+_\\w+) family=CardPrefab persistent=False"

    func handle(game: Game, logLine: LogLine) {
        if logLine.line.contains("rank_window") {
            game.currentGameMode = .ranked
        } else if logLine.line.match(UnloadingCard) {
            let match = logLine.line.matches(UnloadingCard)
            if let match = match.first {
                let cardId = match.value

                if let card = Cards.by(cardId: cardId) {
                    if game.currentMode == .draft && game.previousMode == .hub {
                        Log.verbose?.message("Possible arena card draft : \(card) ?")
                    } else if (game.currentMode == .collectionmanager
                        || game.currentMode == .tavern_brawl)
                        && game.previousMode == .hub {
                        Log.verbose?.message("Possible constructed card draft : \(card) ?")
                    }
                }
            }
        } else if logLine.line.contains("unloading name=Tavern_Brawl") {
            game.currentGameMode = .brawl
        }
    }
}
